File: Samples/DocumentReader-sample/app/src/main/java/com/regula/documentreader/FragmentMain.java
package com.regula.documentreader;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ListView;
import android.widget.TextView;

import com.regula.documentreader.api.DocumentReader;
import com.regula.documentreader.api.completions.IDocumentReaderCompletion;
import com.regula.documentreader.api.completions.IDocumentReaderInitCompletion;
import com.regula.documentreader.api.completions.IDocumentReaderPrepareCompletion;
import com.regula.documentreader.api.results.DocumentReaderScenario;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import static com.regula.documentreader.MainActivity.DO_RFID;
import static com.regula.documentreader.MainActivity.PERMISSIONS_REQUEST_READ_EXTERNAL_STORAGE;
import static com.regula.documentreader.MainActivity.REQUEST_BROWSE_PICTURE;
import static com.regula.documentreader.MainActivity.sharedPreferences;

public class FragmentMain extends Fragment {

    private int selectedPosition;

    private boolean doRfid;

    private TextView showScanner;
    private TextView recognizeImage;

    private CheckBox doRfidCb;

    private ListView scenarioLv;

    private Activity activity;

    private IDocumentReaderCompletion completion;

    private static FragmentMain instance;

    static FragmentMain getInstance(IDocumentReaderCompletion completion){
        if(instance==null){
            instance = new FragmentMain();
        }
        instance.completion = completion;
        return instance;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        activity = (Activity)context;
    }

    @Override
    public void onDetach() {
        super.onDetach();

        activity = null;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_main,container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);


        showScanner = view.findViewById(R.id.showScannerLink);
        recognizeImage = view.findViewById(R.id.recognizeImageLink);

        scenarioLv = view.findViewById(R.id.scenariosList);

        doRfidCb = view.findViewById(R.id.doRfidCb);
    }

    @Override
    public void onResume() {
        super.onResume();

        if(DocumentReader.Instance().getDocumentReaderIsReady()){
            successfulInit();
        }
    }

    void prepareDatabase(Context context, final IDocumentReaderPrepareCompletion completion){
        //preparing database files, it will be downloaded from network only one time and stored on user device
        DocumentReader.Instance().prepareDatabase(context, "Full", new
                IDocumentReaderPrepareCompletion() {

                    @Override
                    public void onPrepareProgressChanged(int i) {
                        completion.onPrepareProgressChanged(i);
                    }

                    @Override
                    public void onPrepareCompleted(boolean b, String s) {
                        completion.onPrepareCompleted(b, s);
                    }
                });
    }

    void init(final Context context,
                 final IDocumentReaderInitCompletion initCompletion){
        try {
            InputStream licInput = getResources().openRawResource(R.raw.regula);
            int available = licInput.available();
            final byte[] license = new byte[available];
            //noinspection ResultOfMethodCallIgnored
            licInput.read(license);

            //Initializing the reader
            DocumentReader.Instance().initializeReader(context, license, new IDocumentReaderInitCompletion() {
                @Override
                public void onInitCompleted(boolean success, String error) {
                    DocumentReader.Instance().customization().edit()
                            .setShowResultStatusMessages(true)
                            .setShowStatusMessages(true);
                    DocumentReader.Instance().functionality().edit()
                            .setVideoCaptureMotionControl(true).apply();

                    //initialization successful
                    if (success) {
                        successfulInit();
                    }

                    initCompletion.onInitCompleted(success, error);
                }
            });

            licInput.close();
        } catch (Exception ex){
            ex.printStackTrace();
        }
    }

    private void successfulInit() {
        showScanner.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //starting video processing
                DocumentReader.Instance().showScanner(getContext(), completion);
            }
        });

        recognizeImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //checking for image browsing permissions
                if (ContextCompat.checkSelfPermission(activity,
                        Manifest.permission.READ_EXTERNAL_STORAGE)
                        != PackageManager.PERMISSION_GRANTED) {

                    ActivityCompat.requestPermissions(activity,
                            new String[]{Manifest.permission.READ_EXTERNAL_STORAGE},
                            PERMISSIONS_REQUEST_READ_EXTERNAL_STORAGE);
                } else {
                    //start image browsing
                    createImageBrowsingRequest();
                }
            }
        });

        if (DocumentReader.Instance().isRFIDAvailableForUse()) {
            //reading shared preferences
            doRfid = sharedPreferences.getBoolean(DO_RFID, false);
            doRfidCb.setChecked(doRfid);
            doRfidCb.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton compoundButton, boolean checked) {
                    doRfid = checked;
                    sharedPreferences.edit().putBoolean(DO_RFID, checked).apply();
                }
            });
        } else {
            doRfidCb.setVisibility(View.GONE);
        }

        //getting current processing scenario and loading available scenarios to ListView
        String currentScenario = DocumentReader.Instance().processParams().scenario;
        ArrayList<String> scenarios = new ArrayList<>();
        for (DocumentReaderScenario scenario : DocumentReader.Instance().availableScenarios) {
            scenarios.add(scenario.name);
        }

        //setting default scenario
        if (currentScenario == null || currentScenario.isEmpty()) {
            currentScenario = scenarios.get(0);
            DocumentReader.Instance().processParams().scenario = currentScenario;
        }

        final ScenarioAdapter adapter =
                new ScenarioAdapter(activity, android.R.layout.simple_list_item_1, scenarios);
        selectedPosition = 0;
        try {
            selectedPosition = adapter.getPosition(currentScenario);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        scenarioLv.setAdapter(adapter);

        scenarioLv.setSelection(selectedPosition);

        scenarioLv.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                //setting selected scenario to DocumentReader params
                DocumentReader.Instance().processParams().scenario = adapter.getItem(i);
                selectedPosition = i;
                adapter.notifyDataSetChanged();

            }
        });
    }

    // creates and starts image browsing intent
    // results will be handled in onActivityResult method
    void createImageBrowsingRequest() {
        Intent intent = new Intent();
        intent.setType("image/*");
        intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true);
        intent.setAction(Intent.ACTION_GET_CONTENT);
        startActivityForResult(Intent.createChooser(intent, "Select Picture"), REQUEST_BROWSE_PICTURE);
    }

    class ScenarioAdapter extends ArrayAdapter<String> {

        public ScenarioAdapter(@NonNull Context context, int resource, @NonNull List<String> objects) {
            super(context, resource, objects);
        }

        @NonNull
        @Override
        public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
            View view = super.getView(position, convertView, parent);

            if(position == selectedPosition){
                view.setBackgroundColor(Color.LTGRAY);
            } else {
                view.setBackgroundColor(Color.TRANSPARENT);
            }
            return view;
        }
    }
}

File: Samples/DocumentReader-sample_kotlin/app/src/main/java/com/regula/documentreader/FragmentMain.kt
package com.regula.documentreader

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.CheckBox
import android.widget.ListView
import android.widget.TextView
import com.regula.documentreader.MainActivity.Companion.sharedPreferences
import com.regula.documentreader.api.DocumentReader
import com.regula.documentreader.api.completions.IDocumentReaderCompletion
import com.regula.documentreader.api.completions.IDocumentReaderInitCompletion
import com.regula.documentreader.api.completions.IDocumentReaderPrepareCompletion
import java.util.*

class FragmentMain : Fragment() {
    private var selectedPosition = 0
    private var doRfid = false
    private var showScanner: TextView? = null
    private var recognizeImage: TextView? = null
    private var doRfidCb: CheckBox? = null
    private var scenarioLv: ListView? = null
    private var activity: Activity? = null
    private var completion: IDocumentReaderCompletion? = null
    override fun onAttach(context: Context) {
        super.onAttach(context)
        activity = context as Activity
    }

    override fun onDetach() {
        super.onDetach()
        activity = null
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_main, container, false)
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        super.onViewCreated(view, savedInstanceState)
        showScanner = view.findViewById(R.id.showScannerLink)
        recognizeImage = view.findViewById(R.id.recognizeImageLink)
        scenarioLv = view.findViewById(R.id.scenariosList)
        doRfidCb = view.findViewById(R.id.doRfidCb)
    }

    override fun onResume() {
        super.onResume()
        if (DocumentReader.Instance().documentReaderIsReady) {
            successfulInit()
        }
    }

    fun prepareDatabase(
        context: Context?,
        completion: IDocumentReaderPrepareCompletion
    ) {
        //preparing database files, it will be downloaded from network only one time and stored on user device
        DocumentReader.Instance()
            .prepareDatabase(context!!, "Full", object : IDocumentReaderPrepareCompletion {
                override fun onPrepareProgressChanged(i: Int) {
                    completion.onPrepareProgressChanged(i)
                }

                override fun onPrepareCompleted(
                    b: Boolean,
                    s: String
                ) {
                    completion.onPrepareCompleted(b, s)
                }
            })
    }

    fun init(
        context: Context?,
        initCompletion: IDocumentReaderInitCompletion
    ) {
        try {
            val licInput = resources.openRawResource(R.raw.regula)
            val available = licInput.available()
            val license = ByteArray(available)
            licInput.read(license)

            //Initializing the reader
            DocumentReader.Instance()
                .initializeReader(context, license) { success, error ->
                    DocumentReader.Instance().customization().edit()
                        .setShowResultStatusMessages(true)
                        .setShowStatusMessages(true)
                    DocumentReader.Instance().functionality().isVideoCaptureMotionControl = true

                    //initialization successful
                    if (success) {
                        successfulInit()
                    }
                    initCompletion.onInitCompleted(success, error)
                }
            licInput.close()
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    private fun successfulInit() {
        showScanner!!.setOnClickListener { //starting video processing
            DocumentReader.Instance().showScanner(context!!, completion!!)
        }
        recognizeImage!!.setOnClickListener {
            //checking for image browsing permissions
            if (ContextCompat.checkSelfPermission(
                    activity!!,
                    Manifest.permission.READ_EXTERNAL_STORAGE
                )
                != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                    activity!!,
                    arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE),
                    MainActivity.PERMISSIONS_REQUEST_READ_EXTERNAL_STORAGE
                )
            } else {
                //start image browsing
                createImageBrowsingRequest()
            }
        }
        if (DocumentReader.Instance().canRFID) {
            //reading shared preferences
            doRfid = sharedPreferences?.getBoolean(MainActivity.DO_RFID, false) ?: false
            doRfidCb!!.isChecked = doRfid
            doRfidCb!!.setOnCheckedChangeListener { compoundButton, checked ->
                doRfid = checked
                sharedPreferences?.edit()?.putBoolean(MainActivity.DO_RFID, checked)
                    ?.apply()
            }
        } else {
            doRfidCb!!.visibility = View.GONE
        }

        //getting current processing scenario and loading available scenarios to ListView
        var currentScenario = DocumentReader.Instance().processParams().scenario
        val scenarios = ArrayList<String>()
        for (scenario in DocumentReader.Instance().availableScenarios) {
            scenarios.add(scenario.name)
        }

        //setting default scenario
        if (currentScenario == null || currentScenario.isEmpty()) {
            currentScenario = scenarios[0]
            DocumentReader.Instance().processParams().scenario = currentScenario
        }
        val adapter =
            ScenarioAdapter(
                activity!!,
                android.R.layout.simple_list_item_1,
                scenarios
            )
        selectedPosition = 0
        try {
            selectedPosition = adapter.getPosition(currentScenario)
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
        scenarioLv!!.adapter = adapter
        scenarioLv!!.setSelection(selectedPosition)
        scenarioLv!!.onItemClickListener =
            OnItemClickListener { adapterView, view, i, l -> //setting selected scenario to DocumentReader params
                DocumentReader.Instance().processParams().scenario = adapter.getItem(i)
                selectedPosition = i
                adapter.notifyDataSetChanged()
            }
    }

    // creates and starts image browsing intent
    // results will be handled in onActivityResult method
    fun createImageBrowsingRequest() {
        val intent = Intent()
        intent.type = "image/*"
        intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
        intent.action = Intent.ACTION_GET_CONTENT
        startActivityForResult(
            Intent.createChooser(intent, "Select Picture"),
            MainActivity.REQUEST_BROWSE_PICTURE
        )
    }

    internal inner class ScenarioAdapter(
        context: Context,
        resource: Int,
        objects: List<String?>
    ) : ArrayAdapter<String?>(context, resource, objects) {
        override fun getView(
            position: Int,
            convertView: View?,
            parent: ViewGroup
        ): View {
            val view = super.getView(position, convertView, parent)
            if (position == selectedPosition) {
                view.setBackgroundColor(Color.LTGRAY)
            } else {
                view.setBackgroundColor(Color.TRANSPARENT)
            }
            return view
        }
    }

    companion object {
        private var instance: FragmentMain? = null
        fun getInstance(completion: IDocumentReaderCompletion?): FragmentMain? {
            if (instance == null) {
                instance = FragmentMain()
            }
            instance!!.completion = completion
            return instance
        }
    }
}